packages/asset-swapper/src/utils/market_operation_utils/path_optimizer.ts

```diff
@@ -1,20 +1,358 @@
+            continue;
+        }
+        // TODO(kimpers): [TKR-241] amounts are sometimes clipped in the router due to precisions loss for number/f64
+        // we can work around it by scaling it and rounding up. However now we end up with a total amount of a couple base units too much
+        const rustInput = BigNumber.min(
+            new BigNumber(routeInput).multipliedBy(scale).integerValue(BigNumber.ROUND_CEIL),
+            input,
+        );
+
+        const current = routeSamplesAndNativeOrders[routeSamplesAndNativeOrders.length - 1];
+        if (!isDexSample(current)) {
+            const nativeFill = nativeOrdersToFills(
+                side,
+                [current],
+                rustInput,
+                opts.outputAmountPerEth,
+                opts.inputAmountPerEth,
+                fees,
+            )[0];
+            // NOTE: For Limit/RFQ orders we are done here. No need to scale output
+            adjustedFills.push(nativeFill);
+            continue;
+        }
+
+        // NOTE: For DexSamples only
+        let fill = createFill(current);
+        const routeSamples = routeSamplesAndNativeOrders as Array<DexSample<FillData>>;
+        // Descend to approach a closer fill for fillData which may not be consistent
+        // throughout the path (UniswapV3) and for a closer guesstimate at
+        // gas used
+        for (let k = routeSamples.length - 1; k >= 0; k--) {
+            if (k === 0) {
+                fill = createFill(routeSamples[0]);
+            }
+            if (rustInput.isGreaterThan(routeSamples[k].input)) {
+                // Between here and the previous fill
+                // HACK: Use the midpoint between the two
+                const left = routeSamples[k];
+                const right = routeSamples[k + 1];
+                if (left && right) {
+                    const leftPrice = left.output.dividedBy(left.input);
+                    const rightPrice = right.output.dividedBy(right.input);
+                    const scaledPrice = leftPrice
+                        .minus(rightPrice)
+                        .dividedBy(left.input.minus(right.input))
+                        .times(rustInput.minus(right.input))
+                        .plus(rightPrice);
+                    const output = scaledPrice
+                        .times(rustInput)
+                        .decimalPlaces(0, side === MarketOperation.Sell ? BigNumber.ROUND_FLOOR : BigNumber.ROUND_CEIL);
+                    fill = createFill({
+                        ...right, // default to the greater (for gas used)
+                        input: rustInput,
+                        output,
+                    });
+                } else {
+                    fill = createFill(left || right);
+                }
+                break;
+            }
+        }
+
+        // HACK: Handle the case where the router can under quote the input
+        // Set the first fill just a tad higher
```

Comment:
Doesn't it make more sense to inflate the last fill? But also clipping it to the max input for that fill? And should it be spread over all the sources? Or just the last? I guess there is no canonical ordering of fills coming from the router, right?

Reply:
The ordering that the router returns is the same as the input order so there's no significance to that. Is there any advantage to doing it on the last fill? (This is another snippet that I blatantly stole), it seems easier to just do it on the first? 🤔 

It would be better to spread this over all the sources but it's probably not worth the effort. The router already tries to do that before returning the results but the `f64/number` precision loss makes it sometimes slightly short anyway.

Once we refactor the router to use a higher precision representation of amounts we should be able to remove this hack entirely!

Reply:
On second thought I wonder if we can just remove this code entirely, I already added [scaling on the input](https://github.com/0xProject/protocol/pull/295/files#diff-899962c6423923a5840eea991afd577aa723139f14440279faffba0b8726ab36R237-R242) before we get here to combat the precision loss. I think we're double adjusting right now 🤔

Reply:
OK I forgot we're creating fills and not samples (which are cumulative). But it still feels like if we're going to be inflating a fill it would be safer using the worst rate (last fill) vs best (first fill).

Reply:
Yeah right, but keep in mind that these amounts stem from rounding/precision errors so they will be in the order of magnitude of a couple thousand wei at the high end 🐛